rfcs/0180-Github-cancel-previous-tasks.md

```diff
@@ -0,0 +1,29 @@
+# RFC 180 - Github cancel previous tasks
+* Comments: [#180](https://github.com/taskcluster/taskcluster-rfcs/pull/180)
+* Proposed by: @lotas
+
+# Summary
+
+Allow to automatically cancel previous pipelines for github push and pull_request events.
+
+## Motivation
+
+This will allow to save computing resources and give more flexibility for the Taskcluster users on Github.
+When several pushes happen to the same branch within a short amount of time, like hotfixes or rebases,
+only the last commit and pipelines matters.
+By cancelling previous task groups and tasks we could save significant amount of redundant tasks.
+
+# Details
+
+This will only apply to non-default branches (and possible protected branches?),
+to allow all pushes to the default branch to have their own builds.
+
+To be flexible and allow some projects to still run checks for all events, we could add new key to the `.taskcluster.yml` configuration.
+We can introduce top-level configuration option `autoCancel: true` (having `false` by default) to control if this behaviour is desired.
+
+Github's webhook handler upon receiving `push` event would check if there are other task groups existing for given branch, that are not HEAD.
+If `autoCancel` is set to `true` it will cancel them (Do we need to cancel all tasks in group one by one or just the parent decision task?)
```

Comment:
If we can catch it fast enough, the decision task will not be able to generate child tasks. But I could imagine some race conditions where the decision task already ran successfully, then we have 60+ child tasks still running. Should keep this in mind going forward.

Reply:
This will always be a bit racey, because in theory any task can create children, so we will always be racing to cancel a task with the task completing. In the case a task dynamically creates a task and does not make it dependent on itself, it may be able to create the task even if the task itself has been cancelled (since workers only routinely check whether they are cancelled, and there are no reasonable means (that wouldn't be a hack) to prevent their credentials from creating new tasks even when the Queue knows they are cancelled, I believe). However, I think we can do cancel all unresolved tasks in the task group in a loop of cancelling all known tasks and then querying the Queue for the tasks in the taskgroup again, until it returns no tasks in scheduled/pending/running state.

In any case, for the most part I think we can almost always guarantee that we cancel all the scheduled/pending/running tasks required, and in the rare case that this approach misses a task or two, it is no big concern, since the goal is to aggressively save resources where we can, and it will make a significant impact on this goal, even if not 100% water proof.

Reply:
Could also create some flag for that task group to not allow any new tasks to be created under it (assuming when any task creates a child task, it's guaranteed to be in the same task group). This may not be as straightforward as I'm thinking it out to be in my head...

Reply:
The idea of task group cancellation came up before in https://github.com/taskcluster/taskcluster-queue/pull/292#issue-367173905 which I coincidentally was looking at today (as it relates to #5948)!

Currently task groups are untameable beasts to some extent - we have no means to restrict their abilities, and they have no metadata associated with them. They are an important concept in taskcluster, but don't really provide and controls around them - this is certainly an area we could look into. I can imagine taskcluster apis for managing task groups, such as preventing them from growing, providing names and descriptions, them having owners, expirations, offering subscription services for notifications, ...